File: backend/python/app/schemas/incident.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.core.incident_sort import IncidentSortField, SortOrder
from app.core.incident_status import IncidentStatus
from app.core.incident_severity import IncidentSeverity


class IncidentCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    title: str = Field(min_length=1, max_length=200)
    description: str | None = None
    status: IncidentStatus
    severity: IncidentSeverity
    service_id: int = Field(gt=0)
    started_at: datetime
    resolved_at: datetime | None = None


class IncidentUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    title: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = None
    status: IncidentStatus | None = None
    severity: IncidentSeverity | None = None
    service_id: int | None = Field(default=None, gt=0)
    assigned_to_id: int | None = None
    started_at: datetime | None = None
    resolved_at: datetime | None = None


class IncidentResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    title: str
    description: str | None
    status: str
    severity: str
    service_id: int
    created_by_id: int
    assigned_to_id: int | None
    started_at: datetime
    resolved_at: datetime | None
    created_at: datetime
    updated_at: datetime


class IncidentListQuery(BaseModel):
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=20, ge=1, le=100)
    search: str | None = None
    status: IncidentStatus | None = None
    severity: IncidentSeverity | None = None
    assigned_to_id: int | None = None
    service_id: int | None = None
    sort_by: IncidentSortField | None = None
    sort_order: SortOrder | None = None


class PaginatedIncidentResponse(BaseModel):
    items: list[IncidentResponse]
    page: int
    page_size: int
    total: int
    total_pages: int

File: frontend/src/types/api/incident.ts
export type ApiIncidentStatus = 'investigating' | 'identified' | 'monitoring' | 'resolved'

export type ApiIncidentSeverity = 'critical' | 'high' | 'medium' | 'low'

export type ApiIncidentSortField =
  | 'started_at'
  | 'created_at'
  | 'updated_at'
  | 'severity'
  | 'status'

export type ApiSortOrder = 'asc' | 'desc'

export interface ApiIncident {
  id: number
  title: string
  description: string | null
  status: string
  severity: string
  service_id: number
  created_by_id: number
  assigned_to_id: number | null
  started_at: string
  resolved_at: string | null
  created_at: string
  updated_at: string
}

export interface PaginatedIncidents {
  items: ApiIncident[]
  page: number
  page_size: number
  total: number
  total_pages: number
}

export interface ApiIncidentEvent {
  id: number
  incident_id: number
  author_id: number
  event_type: string
  message: string
  created_at: string
}

export interface ApiIncidentComment {
  id: number
  incident_id: number
  author_id: number
  content: string
  created_at: string
  updated_at: string
}

export interface IncidentListParams {
  page?: number
  page_size?: number
  search?: string
  status?: ApiIncidentStatus
  severity?: ApiIncidentSeverity
  service_id?: number
  assigned_to_id?: number
  sort_by?: ApiIncidentSortField
  sort_order?: ApiSortOrder
}

/** POST /api/v1/incidents — backend sets status, creator, and assignee. */
export interface ApiIncidentCreate {
  title: string
  description: string | null
  severity: ApiIncidentSeverity
  service_id: number
}

/** PATCH /api/v1/incidents/{id} — maps to `IncidentUpdate`. */
export interface ApiIncidentUpdate {
  title?: string | null
  description?: string | null
  status?: ApiIncidentStatus | null
  severity?: ApiIncidentSeverity | null
  service_id?: number | null
  assigned_to_id?: number | null
  started_at?: string | null
  resolved_at?: string | null
}

/** POST /api/v1/incidents/{id}/events */
export interface ApiIncidentEventCreate {
  event_type: string
  message: string
}

/** POST /api/v1/incidents/{id}/comments — maps to `IncidentCommentCreateBody`. */
export interface ApiIncidentCommentCreate {
  content: string
}

/** PATCH /api/v1/incidents/{id}/comments/{comment_id} — maps to `IncidentCommentUpdate`. */
export interface ApiIncidentCommentUpdate {
  content: string
}
